### PR TITLE
Kill GPG daemons

### PR DESCRIPTION
**Pull request by dbnicholson**
GnuPG 2 has a fun implementation where it spawns a `gpg-agent` daemon that never exits for any operation. Coupled with libostree's usage of temporary directories for keyring construction, this can leave a lot of `gpg-agent`s running. When we know that the components are associated with a temporary directory that's being deleted, we can call `gpgconf --kill all` to get rid of them.

Now when I run the test suite there are no `gpg-agent`s left behind!

Closes: #727

### CODE REVIEW

**Comment by dbnicholson** [comment]
> No objection to this, but I think probably a better fix to teach gpgme upstream to use `prctl(PR_SET_PDEATHSIG, SIGTERM)`.

If only it worked that way... The helper in gnupg intentionally starts the agent in daemon mode where it forks and the parent exits. The idea is that in normal operation you do want the gpg-agent to be long lived since your GPG home directory tends to be (very) persistent. gpgme doesn't control that. You could still set that, but it would more or less be when the user session ended, which might be a long time from now.

File: src/libotutil/ot-gpg-utils.c [hunk 1]
```diff
@@ -437,3 +437,30 @@ ot_gpgme_new_ctx (const char *homedir,
...
+  g_return_if_fail (homedir != NULL);
+
+  /* Run gpgconf --kill all with GNUPGHOME set to the homedir */
+  const char *gpgconf = gpgme_get_dirinfo ("gpgconf-name");
+  if (gpgconf == NULL) {
+    g_warning ("Could not get name for gpgconf tool");
+    return;
+  }
+  g_autoptr(GSubprocessLauncher) launcher = g_subprocess_launcher_new(G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_setenv(launcher, "GNUPGHOME", homedir, TRUE);
```
[the remark this answers is not in the document]
Comment (dbnicholson) [reply src/libotutil/ot-gpg-utils.c]: It's a hidden option that was added in gnupg-2.1.13. It doesn't seem that gpgme has any required gnupg minimum version, so it's probably safest to use the environment variable.

**Comment by cgwalters** [comment]
> The idea is that in normal operation you do want the gpg-agent to be long lived since your GPG home directory tends to be (very) persistent. 

Ah yeah, then we'd also want to add an option to disable that daemonization.

**Comment by cgwalters** [comment]
One thing I'd note is this isn't just about the test suite - it also applies to the runtime verification.  Which is  a good thing, but does increase the risk/scope.

Hm, looks like Travis builds fell over with e.g.

```
Component not found

(ostree remote add:4083): OSTree-WARNING **: Killing GPG components with gpgconf failed: Child process exited with code 1
/home/user/ostree/tests/test-pull-collections.sh: line 94:  4083 Trace/breakpoint trap   (core dumped) ${CMD_PREFIX} ostree "--repo=${repo}" remote add "${remote_repo}-remote" "file://$(pwd)/${remote_repo}" "$@" --gpg-import="${test_tmpdir}/gpghome/key1.asc"

```

**Comment by cgwalters** [comment]
This came up in another PR I'm too lazy to find but I don't like `g_warning()`.  Let's just drop them which may also fix the test suite which uses `G_DEBUG=fatal-warnings` I think.

**Comment by dbnicholson** [comment]
I decided to rework this with a few changes after a second look. Summary of changes:

* Rather than calling `gpgconf --kill all`, use the classic `gpg-connect-agent killagent /bye` as is done in libtest. I'd forgotten about that and we can be pretty sure it will work as advertised as that's been the "correct" way to kill a running gpg-agent since the gnupg1 days.

* Instead of inventing a new way in tests to kill the agent, use the existing `libtest_cleanup_gpg()` but make it always run as an `EXIT` hook.

I also changed the 2 `g_warning`s to `g_debug`s as requested.

As it turns out, gnupg does now do what you'd want - watch the homedir with inotify and exit when it's deleted. Unfortunately, that's only available somewhat recently - https://dev.gnupg.org/T2756.

**Comment by cgwalters** [comment]
Thanks!

@rh-atomic-bot r+ 513633

**Comment by cgwalters** [comment]
@rh-atomic-bot r+ 6139633

**Comment by cgwalters** [comment]
@rh-atomic-bot r+ 5139633

**Pull request closed without merging**